core/src/main/scala/org/http4s/internal/package.scala

```diff
@@ -14,37 +14,40 @@ import java.util.concurrent.{
 }
 
 import cats.effect.implicits._
-import cats.effect.{Async, Concurrent, ConcurrentEffect, ContextShift, Effect, IO}
+import cats.effect.std.Dispatcher
+import cats.effect.kernel.{Async, Sync}
 import cats.implicits._
 import fs2.{Chunk, Pipe, Pull, RaiseThrowable, Stream}
 import java.nio.{ByteBuffer, CharBuffer}
-import org.http4s.util.execution.direct
 import org.log4s.Logger
 
 import scala.concurrent.{ExecutionContext, Future}
 import scala.util.control.NoStackTrace
-import scala.util.{Failure, Success}
 import java.nio.charset.MalformedInputException
 import java.nio.charset.UnmappableCharacterException
 
 package object internal {
   // Like fs2.async.unsafeRunAsync before 1.0.  Convenient for when we
   // have an ExecutionContext but not a Timer.
-  private[http4s] def unsafeRunAsync[F[_], A](fa: F[A])(
-      f: Either[Throwable, A] => IO[Unit])(implicit F: Effect[F], ec: ExecutionContext): Unit =
-    F.runAsync(Async.shift(ec) *> fa)(f).unsafeRunSync()
+  private[http4s] def unsafeRunAsync[F[_]: Async, A](fa: F[A])(
+      f: Either[Throwable, A] => F[Unit])(implicit
+      dispatcher: Dispatcher[F],
+      ec: ExecutionContext): Unit =
+    dispatcher.unsafeRunSync(fa.evalOn(ec).attemptTap(f).void)
 
-  private[http4s] def loggingAsyncCallback[A](logger: Logger)(
-      attempt: Either[Throwable, A]): IO[Unit] =
+  private[http4s] def loggingAsyncCallback[F[_], A](logger: Logger)(attempt: Either[Throwable, A])(
+      implicit F: Sync[F]): F[Unit] =
     attempt match {
-      case Left(e) => IO(logger.error(e)("Error in asynchronous callback"))
-      case Right(_) => IO.unit
+      case Left(e) => F.delay(logger.error(e)("Error in asynchronous callback"))
+      case Right(_) => F.unit
     }
 
   // Inspired by https://github.com/functional-streams-for-scala/fs2/blob/14d20f6f259d04df410dc3b1046bc843a19d73e5/io/src/main/scala/fs2/io/io.scala#L140-L141
-  private[http4s] def invokeCallback[F[_]](logger: Logger)(f: => Unit)(implicit
-      F: ConcurrentEffect[F]): Unit =
-    F.runAsync(F.start(F.delay(f)).flatMap(_.join))(loggingAsyncCallback(logger)).unsafeRunSync()
+  private[http4s] def invokeCallback[F[_]](logger: Logger)(
+      f: => Unit)(implicit F: Async[F], dispatcher: Dispatcher[F]): Unit =
```

Comment:
I wouldn't generally recommend making the `Dispatcher` implicit. You *can*, but it's a bit weird given that it needs to live within a `Resource`.